File: frontend/lib/sheets.js
import { google } from 'googleapis';

let sheetsClient = null;

export function getSheetsClient() {
  if (sheetsClient) return { srv: sheetsClient, spreadsheetId: process.env.SPREADSHEET_ID };

  let credJSON = process.env.GOOGLE_CREDENTIALS_JSON;
  if (!credJSON) {
    const c1 = process.env.GOOGLE_CRED_1 || '';
    const c2 = process.env.GOOGLE_CRED_2 || '';
    const c3 = process.env.GOOGLE_CRED_3 || '';
    credJSON = c1 + c2 + c3;
  }

  if (!credJSON) {
    throw new Error('GOOGLE_CREDENTIALS_JSON tidak dikonfigurasi di Environment Variables.');
  }

  const credentials = JSON.parse(credJSON);
  const auth = new google.auth.GoogleAuth({
    credentials,
    scopes: ['https://www.googleapis.com/auth/spreadsheets'],
  });

  sheetsClient = google.sheets({ version: 'v4', auth });
  return { srv: sheetsClient, spreadsheetId: process.env.SPREADSHEET_ID };
}

function safeString(row, idx) {
  if (!row || idx >= row.length || row[idx] === null || row[idx] === undefined) return '';
  return String(row[idx]);
}

function safeInt(row, idx) {
  const val = safeString(row, idx);
  if (!val) return 0;
  const num = parseInt(val, 10);
  return isNaN(num) ? 0 : num;
}

// ============ PESANAN ============

export async function getAllPesanan() {
  const { srv, spreadsheetId } = getSheetsClient();
  const res = await srv.spreadsheets.values.get({
    spreadsheetId,
    range: 'Pesanan!A2:K',
  });

  const rows = res.data.values || [];
  return rows.map((row) => {
    let items = [];
    const itemsJSON = safeString(row, 6);
    if (itemsJSON) {
      try { items = JSON.parse(itemsJSON); } catch (e) {}
    }

    return {
      id: safeString(row, 0),
      nama_pengantin: safeString(row, 1),
      alamat: safeString(row, 2),
      no_hp: safeString(row, 3),
      tanggal_pasang: safeString(row, 4),
      tanggal_bongkar: safeString(row, 5),
      items,
      total_harga: safeInt(row, 7),
      status: safeString(row, 8),
      catatan: safeString(row, 9),
      created_at: safeString(row, 10),
    };
  });
}

export async function getPesananById(id) {
  const list = await getAllPesanan();
  const idx = list.findIndex((p) => p.id === id);
  if (idx === -1) return null;
  return { data: list[idx], rowNum: idx + 2 };
}

export async function createPesanan(pesanan) {
  const { srv, spreadsheetId } = getSheetsClient();
  const itemsJSON = JSON.stringify(pesanan.items || []);
  const now = new Date().toISOString();

  const values = [
    pesanan.id || `ORD-${Date.now()}`,
    pesanan.nama_pengantin || '',
    pesanan.alamat || '',
    pesanan.no_hp || '',
    pesanan.tanggal_pasang || '',
    pesanan.tanggal_bongkar || '',
    itemsJSON,
    pesanan.total_harga || 0,
    pesanan.status || 'Pending',
    pesanan.catatan || '',
    now,
  ];

  await srv.spreadsheets.values.append({
    spreadsheetId,
    range: 'Pesanan!A:K',
    valueInputOption: 'RAW',
    requestBody: { values: [values] },
  });

  return { ...pesanan, id: values[0], created_at: now };
}

export async function updatePesanan(id, pesanan) {
  const existing = await getPesananById(id);
  if (!existing) throw new Error('Pesanan tidak ditemukan');

  const { srv, spreadsheetId } = getSheetsClient();
  const itemsJSON = JSON.stringify(pesanan.items || []);

  const values = [
    id,
    pesanan.nama_pengantin || '',
    pesanan.alamat || '',
    pesanan.no_hp || '',
    pesanan.tanggal_pasang || '',
    pesanan.tanggal_bongkar || '',
    itemsJSON,
    pesanan.total_harga || 0,
    pesanan.status || 'Pending',
    pesanan.catatan || '',
    existing.data.created_at || new Date().toISOString(),
  ];

  const range = `Pesanan!A${existing.rowNum}:K${existing.rowNum}`;
  await srv.spreadsheets.values.update({
    spreadsheetId,
    range,
    valueInputOption: 'RAW',
    requestBody: { values: [values] },
  });

  return { ...pesanan, id };
}

export async function deletePesanan(id) {
  const existing = await getPesananById(id);
  if (!existing) throw new Error('Pesanan tidak ditemukan');

  const { srv, spreadsheetId } = getSheetsClient();
  const sp = await srv.spreadsheets.get({ spreadsheetId });
  const sheet = sp.data.sheets.find((s) => s.properties.title === 'Pesanan');
  const sheetId = sheet?.properties?.sheetId || 0;

  await srv.spreadsheets.batchUpdate({
    spreadsheetId,
    requestBody: {
      requests: [
        {
          deleteDimension: {
            range: {
              sheetId,
              dimension: 'ROWS',
              startIndex: existing.rowNum - 1,
              endIndex: existing.rowNum,
            },
          },
        },
      ],
    },
  });
}

// ============ ITEMS ============

export async function getAllItems() {
  const { srv, spreadsheetId } = getSheetsClient();
  const res = await srv.spreadsheets.values.get({
    spreadsheetId,
    range: 'Items!A2:E',
  });

  const rows = res.data.values || [];
  return rows.map((row) => ({
    id: safeString(row, 0),
    nama: safeString(row, 1),
    harga: safeInt(row, 2),
    gambar_url: safeString(row, 3),
    deskripsi: safeString(row, 4),
  }));
}

export async function getItemById(id) {
  const items = await getAllItems();
  const idx = items.findIndex((i) => i.id === id);
  if (idx === -1) return null;
  return { data: items[idx], rowNum: idx + 2 };
}

export async function createItem(item) {
  const { srv, spreadsheetId } = getSheetsClient();
  const id = item.id || `ITM-${Date.now()}`;

  const values = [
    id,
    item.nama || '',
    item.harga || 0,
    item.gambar_url || '',
    item.deskripsi || '',
  ];

  await srv.spreadsheets.values.append({
    spreadsheetId,
    range: 'Items!A:E',
    valueInputOption: 'RAW',
    requestBody: { values: [values] },
  });

  return { ...item, id };
}

export async function updateItem(id, item) {
  const existing = await getItemById(id);
  if (!existing) throw new Error('Item tidak ditemukan');

  const { srv, spreadsheetId } = getSheetsClient();
  const values = [
    id,
    item.nama || '',
    item.harga || 0,
    item.gambar_url || '',
    item.deskripsi || '',
  ];

  const range = `Items!A${existing.rowNum}:E${existing.rowNum}`;
  await srv.spreadsheets.values.update({
    spreadsheetId,
    range,
    valueInputOption: 'RAW',
    requestBody: { values: [values] },
  });

  return { ...item, id };
}

export async function deleteItem(id) {
  const existing = await getItemById(id);
  if (!existing) throw new Error('Item tidak ditemukan');

  const { srv, spreadsheetId } = getSheetsClient();
  const sp = await srv.spreadsheets.get({ spreadsheetId });
  const sheet = sp.data.sheets.find((s) => s.properties.title === 'Items');
  const sheetId = sheet?.properties?.sheetId || 0;

  await srv.spreadsheets.batchUpdate({
    spreadsheetId,
    requestBody: {
      requests: [
        {
          deleteDimension: {
            range: {
              sheetId,
              dimension: 'ROWS',
              startIndex: existing.rowNum - 1,
              endIndex: existing.rowNum,
            },
          },
        },
      ],
    },
  });
}

// ============ MODEL DEKORASI ============

export async function getAllModels() {
  const { srv, spreadsheetId } = getSheetsClient();
  const res = await srv.spreadsheets.values.get({
    spreadsheetId,
    range: 'Model!A2:E',
  });

  const rows = res.data.values || [];
  return rows.map((row) => ({
    id: safeString(row, 0),
    nama: safeString(row, 1),
    deskripsi: safeString(row, 2),
    gambar_url: safeString(row, 3),
  }));
}

export async function getModelById(id) {
  const models = await getAllModels();
  const idx = models.findIndex((m) => m.id === id);
  if (idx === -1) return null;
  return { data: models[idx], rowNum: idx + 2 };
}

export async function createModel(model) {
  const { srv, spreadsheetId } = getSheetsClient();
  const id = model.id || `MDL-${Date.now()}`;

  const values = [
    id,
    model.nama || '',
    model.deskripsi || '',
    model.gambar_url || '',
  ];

  await srv.spreadsheets.values.append({
    spreadsheetId,
    range: 'Model!A:E',
    valueInputOption: 'RAW',
    requestBody: { values: [values] },
  });

  return { ...model, id };
}

export async function updateModel(id, model) {
  const existing = await getModelById(id);
  if (!existing) throw new Error('Model tidak ditemukan');

  const { srv, spreadsheetId } = getSheetsClient();
  const values = [
    id,
    model.nama || '',
    model.deskripsi || '',
    model.gambar_url || '',
  ];

  const range = `Model!A${existing.rowNum}:D${existing.rowNum}`;
  await srv.spreadsheets.values.update({
    spreadsheetId,
    range,
    valueInputOption: 'RAW',
    requestBody: { values: [values] },
  });

  return { ...model, id };
}

export async function deleteModel(id) {
  const existing = await getModelById(id);
  if (!existing) throw new Error('Model tidak ditemukan');

  const { srv, spreadsheetId } = getSheetsClient();
  const sp = await srv.spreadsheets.get({ spreadsheetId });
  const sheet = sp.data.sheets.find((s) => s.properties.title === 'Model');
  const sheetId = sheet?.properties?.sheetId || 0;

  await srv.spreadsheets.batchUpdate({
    spreadsheetId,
    requestBody: {
      requests: [
        {
          deleteDimension: {
            range: {
              sheetId,
              dimension: 'ROWS',
              startIndex: existing.rowNum - 1,
              endIndex: existing.rowNum,
            },
          },
        },
      ],
    },
  });
}

File: backend/services/sheets.go
package services

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"strconv"
	"time"

	"luwes-dekorasi-api/models"

	"google.golang.org/api/option"
	"google.golang.org/api/sheets/v4"
)

type SheetsService struct {
	srv           *sheets.Service
	spreadsheetID string
}

func NewSheetsService() (*SheetsService, error) {
	ctx := context.Background()

	credJSON := os.Getenv("GOOGLE_CREDENTIALS_JSON")
	if credJSON == "" {
		// Bypass limit panjang karakter Back4App (1023 char limit)
		cred1 := os.Getenv("GOOGLE_CRED_1")
		cred2 := os.Getenv("GOOGLE_CRED_2")
		cred3 := os.Getenv("GOOGLE_CRED_3")
		credJSON = cred1 + cred2 + cred3
	}

	if credJSON == "" {
		return nil, fmt.Errorf("GOOGLE_CREDENTIALS_JSON not set")
	}

	srv, err := sheets.NewService(ctx, option.WithCredentialsJSON([]byte(credJSON)))
	if err != nil {
		return nil, fmt.Errorf("failed to create sheets service: %w", err)
	}

	spreadsheetID := os.Getenv("SPREADSHEET_ID")
	if spreadsheetID == "" {
		return nil, fmt.Errorf("SPREADSHEET_ID not set")
	}

	return &SheetsService{srv: srv, spreadsheetID: spreadsheetID}, nil
}

// ============ PESANAN ============

func (s *SheetsService) GetAllPesanan() ([]models.Pesanan, error) {
	resp, err := s.srv.Spreadsheets.Values.Get(s.spreadsheetID, "Pesanan!A2:K").Do()
	if err != nil {
		return nil, err
	}

	var pesananList []models.Pesanan
	for _, row := range resp.Values {
		p := models.Pesanan{
			ID:             safeString(row, 0),
			NamaPengantin:  safeString(row, 1),
			Alamat:         safeString(row, 2),
			NoHP:           safeString(row, 3),
			TanggalPasang:  safeString(row, 4),
			TanggalBongkar: safeString(row, 5),
			TotalHarga:     safeInt64(row, 7),
			Status:         safeString(row, 8),
			Catatan:        safeString(row, 9),
		}

		itemsJSON := safeString(row, 6)
		if itemsJSON != "" {
			json.Unmarshal([]byte(itemsJSON), &p.Items)
		}

		createdStr := safeString(row, 10)
		if createdStr != "" {
			p.CreatedAt, _ = time.Parse(time.RFC3339, createdStr)
		}

		pesananList = append(pesananList, p)
	}

	return pesananList, nil
}

func (s *SheetsService) GetPesananByID(id string) (*models.Pesanan, int, error) {
	pesananList, err := s.GetAllPesanan()
	if err != nil {
		return nil, 0, err
	}

	for i, p := range pesananList {
		if p.ID == id {
			return &p, i + 2, nil // +2 because row 1 is header, rows are 1-indexed
		}
	}

	return nil, 0, fmt.Errorf("pesanan not found")
}

func (s *SheetsService) CreatePesanan(p models.Pesanan) error {
	itemsJSON, _ := json.Marshal(p.Items)

	values := []interface{}{
		p.ID,
		p.NamaPengantin,
		p.Alamat,
		p.NoHP,
		p.TanggalPasang,
		p.TanggalBongkar,
		string(itemsJSON),
		p.TotalHarga,
		p.Status,
		p.Catatan,
		p.CreatedAt.Format(time.RFC3339),
	}

	vr := &sheets.ValueRange{
		Values: [][]interface{}{values},
	}

	_, err := s.srv.Spreadsheets.Values.Append(
		s.spreadsheetID, "Pesanan!A:K", vr,
	).ValueInputOption("RAW").Do()

	return err
}

func (s *SheetsService) UpdatePesanan(id string, p models.Pesanan) error {
	_, rowNum, err := s.GetPesananByID(id)
	if err != nil {
		return err
	}

	itemsJSON, _ := json.Marshal(p.Items)

	values := []interface{}{
		p.ID,
		p.NamaPengantin,
		p.Alamat,
		p.NoHP,
		p.TanggalPasang,
		p.TanggalBongkar,
		string(itemsJSON),
		p.TotalHarga,
		p.Status,
		p.Catatan,
		p.CreatedAt.Format(time.RFC3339),
	}

	vr := &sheets.ValueRange{
		Values: [][]interface{}{values},
	}

	rangeStr := fmt.Sprintf("Pesanan!A%d:K%d", rowNum, rowNum)
	_, err = s.srv.Spreadsheets.Values.Update(
		s.spreadsheetID, rangeStr, vr,
	).ValueInputOption("RAW").Do()

	return err
}

func (s *SheetsService) DeletePesanan(id string) error {
	_, rowNum, err := s.GetPesananByID(id)
	if err != nil {
		return err
	}

	// Get spreadsheet to find sheet ID
	sp, err := s.srv.Spreadsheets.Get(s.spreadsheetID).Do()
	if err != nil {
		return err
	}

	var sheetID int64
	for _, sheet := range sp.Sheets {
		if sheet.Properties.Title == "Pesanan" {
			sheetID = sheet.Properties.SheetId
			break
		}
	}

	req := &sheets.BatchUpdateSpreadsheetRequest{
		Requests: []*sheets.Request{
			{
				DeleteDimension: &sheets.DeleteDimensionRequest{
					Range: &sheets.DimensionRange{
						SheetId:    sheetID,
						Dimension:  "ROWS",
						StartIndex: int64(rowNum - 1),
						EndIndex:   int64(rowNum),
					},
				},
			},
		},
	}

	_, err = s.srv.Spreadsheets.BatchUpdate(s.spreadsheetID, req).Do()
	return err
}

// ============ ITEMS ============

func (s *SheetsService) GetAllItems() ([]models.Item, error) {
	resp, err := s.srv.Spreadsheets.Values.Get(s.spreadsheetID, "Items!A2:E").Do()
	if err != nil {
		return nil, err
	}

	var items []models.Item
	for _, row := range resp.Values {
		item := models.Item{
			ID:        safeString(row, 0),
			Nama:      safeString(row, 1),
			Harga:     safeInt64(row, 2),
			GambarURL: safeString(row, 3),
			Deskripsi: safeString(row, 4),
		}
		items = append(items, item)
	}

	return items, nil
}

func (s *SheetsService) GetItemByID(id string) (*models.Item, int, error) {
	items, err := s.GetAllItems()
	if err != nil {
		return nil, 0, err
	}

	for i, item := range items {
		if item.ID == id {
			return &item, i + 2, nil
		}
	}

	return nil, 0, fmt.Errorf("item not found")
}

func (s *SheetsService) CreateItem(item models.Item) error {
	values := []interface{}{
		item.ID,
		item.Nama,
		item.Harga,
		item.GambarURL,
		item.Deskripsi,
	}

	vr := &sheets.ValueRange{
		Values: [][]interface{}{values},
	}

	_, err := s.srv.Spreadsheets.Values.Append(
		s.spreadsheetID, "Items!A:E", vr,
	).ValueInputOption("RAW").Do()

	return err
}

func (s *SheetsService) UpdateItem(id string, item models.Item) error {
	_, rowNum, err := s.GetItemByID(id)
	if err != nil {
		return err
	}

	values := []interface{}{
		item.ID,
		item.Nama,
		item.Harga,
		item.GambarURL,
		item.Deskripsi,
	}

	vr := &sheets.ValueRange{
		Values: [][]interface{}{values},
	}

	rangeStr := fmt.Sprintf("Items!A%d:E%d", rowNum, rowNum)
	_, err = s.srv.Spreadsheets.Values.Update(
		s.spreadsheetID, rangeStr, vr,
	).ValueInputOption("RAW").Do()

	return err
}

func (s *SheetsService) DeleteItem(id string) error {
	_, rowNum, err := s.GetItemByID(id)
	if err != nil {
		return err
	}

	sp, err := s.srv.Spreadsheets.Get(s.spreadsheetID).Do()
	if err != nil {
		return err
	}

	var sheetID int64
	for _, sheet := range sp.Sheets {
		if sheet.Properties.Title == "Items" {
			sheetID = sheet.Properties.SheetId
			break
		}
	}

	req := &sheets.BatchUpdateSpreadsheetRequest{
		Requests: []*sheets.Request{
			{
				DeleteDimension: &sheets.DeleteDimensionRequest{
					Range: &sheets.DimensionRange{
						SheetId:    sheetID,
						Dimension:  "ROWS",
						StartIndex: int64(rowNum - 1),
						EndIndex:   int64(rowNum),
					},
				},
			},
		},
	}

	_, err = s.srv.Spreadsheets.BatchUpdate(s.spreadsheetID, req).Do()
	return err
}

// ============ USERS ============

func (s *SheetsService) GetUserByUsername(username string) (*models.User, error) {
	resp, err := s.srv.Spreadsheets.Values.Get(s.spreadsheetID, "Users!A2:E").Do()
	if err != nil {
		return nil, err
	}

	for _, row := range resp.Values {
		if safeString(row, 1) == username {
			return &models.User{
				ID:           safeString(row, 0),
				Username:     safeString(row, 1),
				PasswordHash: safeString(row, 2),
				Nama:         safeString(row, 3),
				Role:         safeString(row, 4),
			}, nil
		}
	}

	return nil, fmt.Errorf("user not found")
}

// ==========================================
// MODEL DEKORASI CRUD
// ==========================================

func (s *SheetsService) GetAllModels() ([]models.ModelDekor, error) {
	readRange := "Model!A2:E"
	resp, err := s.srv.Spreadsheets.Values.Get(s.spreadsheetID, readRange).Do()
	if err != nil {
		return nil, err
	}

	var data []models.ModelDekor
	for _, row := range resp.Values {
		if len(row) < 2 {
			continue // ID dan Nama minimal ada
		}
		data = append(data, models.ModelDekor{
			ID:        safeString(row, 0),
			Nama:      safeString(row, 1),
			Deskripsi: safeString(row, 2),
			GambarURL: safeString(row, 3),
		})
	}
	return data, nil
}

func (s *SheetsService) GetModelByID(id string) (*models.ModelDekor, int, error) {
	modelsList, err := s.GetAllModels()
	if err != nil {
		return nil, 0, err
	}

	for i, m := range modelsList {
		if m.ID == id {
			return &m, i + 2, nil // +2 karena header dan 0-index
		}
	}
	return nil, 0, fmt.Errorf("model not found")
}

func (s *SheetsService) CreateModel(model models.ModelDekor) error {
	writeRange := "Model!A:E"
	var vr sheets.ValueRange
	vr.Values = append(vr.Values, []interface{}{
		model.ID,
		model.Nama,
		model.Deskripsi,
		model.GambarURL,
	})

	_, err := s.srv.Spreadsheets.Values.Append(s.spreadsheetID, writeRange, &vr).
		ValueInputOption("RAW").Do()
	return err
}

func (s *SheetsService) UpdateModel(id string, model models.ModelDekor) error {
	_, rowIdx, err := s.GetModelByID(id)
	if err != nil {
		return err
	}

	updateRange := fmt.Sprintf("Model!A%d:D%d", rowIdx, rowIdx)
	var vr sheets.ValueRange
	vr.Values = append(vr.Values, []interface{}{
		model.ID,
		model.Nama,
		model.Deskripsi,
		model.GambarURL,
	})

	_, err = s.srv.Spreadsheets.Values.Update(s.spreadsheetID, updateRange, &vr).
		ValueInputOption("RAW").Do()
	return err
}

func (s *SheetsService) DeleteModel(id string) error {
	_, rowIdx, err := s.GetModelByID(id)
	if err != nil {
		return err
	}

	sp, err := s.srv.Spreadsheets.Get(s.spreadsheetID).Do()
	if err != nil {
		return err
	}

	var sheetID int64
	for _, sheet := range sp.Sheets {
		if sheet.Properties.Title == "Model" {
			sheetID = sheet.Properties.SheetId
			break
		}
	}

	req := &sheets.BatchUpdateSpreadsheetRequest{
		Requests: []*sheets.Request{
			{
				DeleteDimension: &sheets.DeleteDimensionRequest{
					Range: &sheets.DimensionRange{
						SheetId:    sheetID, 
						Dimension:  "ROWS",
						StartIndex: int64(rowIdx - 1),
						EndIndex:   int64(rowIdx),
					},
				},
			},
		},
	}

	_, err = s.srv.Spreadsheets.BatchUpdate(s.spreadsheetID, req).Do()
	return err
}

// ============ HELPERS ============

func safeString(row []interface{}, idx int) string {
	if idx >= len(row) {
		return ""
	}
	return fmt.Sprintf("%v", row[idx])
}

func safeInt64(row []interface{}, idx int) int64 {
	s := safeString(row, idx)
	if s == "" {
		return 0
	}
	v, _ := strconv.ParseInt(s, 10, 64)
	return v
}
